File: js/competitive-binding.js
// competitive-binding.js — Pillar 2: 실제 Scaled Dot-Product Multi-Head Attention
// =================================================================================
// 핵심 변경: Dense 레이어 → 진짜 Attention 메커니즘
//   Attention(Q, K, V) = softmax(Q·K^T / √d_k) · V
//   Multi-Head: 4 헤드 → concat → linear projection
// =================================================================================

class CompetitiveBinding {
    constructor() {
        // 아키텍처 파라미터
        this.moleculeDim = 32;     // 분자 임베딩 차원
        this.numHeads = 4;
        this.headDim = 8;          // d_k = moleculeDim / numHeads
        this.numReceptors = 8;

        // 학습 가능 가중치
        this.attentionWeights = null;
        this.receptorWeights = null;
        this.projectionModel = null;
        this.trained = false;

        // 가상 수용체 정의 (특화 영역)
        this.receptorProfiles = [
            { name: 'OR1: 플로럴', sensitivity: ['floral', 'rose', 'jasmine', 'violet'] },
            { name: 'OR2: 시트러스', sensitivity: ['citrus', 'lemon', 'orange', 'bergamot'] },
            { name: 'OR3: 우디', sensitivity: ['woody', 'cedar', 'sandalwood', 'vetiver'] },
            { name: 'OR4: 스파이시', sensitivity: ['spicy', 'clove', 'cinnamon', 'pepper'] },
            { name: 'OR5: 스위트', sensitivity: ['sweet', 'vanilla', 'caramel', 'honey'] },
            { name: 'OR6: 프레시', sensitivity: ['fresh', 'clean', 'aquatic', 'ozonic'] },
            { name: 'OR7: 그린', sensitivity: ['green', 'leafy', 'herbal', 'grassy'] },
            { name: 'OR8: 머스크', sensitivity: ['musk', 'amber', 'warm', 'powdery'] }
        ];
    }

    // ==========================================================
    // Attention 가중치 초기화
    // ==========================================================
    initAttentionWeights() {
        const d = this.moleculeDim;
        const h = this.numHeads;
        const dk = this.headDim;

        this.attentionWeights = {};

        // 각 헤드의 Q, K, V 프로젝션
        for (let i = 0; i < h; i++) {
            this.attentionWeights[`W_Q_${i}`] = tf.variable(
                tf.randomNormal([d, dk]).mul(tf.scalar(Math.sqrt(2 / d)))
            );
            this.attentionWeights[`W_K_${i}`] = tf.variable(
                tf.randomNormal([d, dk]).mul(tf.scalar(Math.sqrt(2 / d)))
            );
            this.attentionWeights[`W_V_${i}`] = tf.variable(
                tf.randomNormal([d, dk]).mul(tf.scalar(Math.sqrt(2 / d)))
            );
        }

        // 멀티헤드 출력 프로젝션: [numHeads * dk] → [d]
        this.attentionWeights['W_O'] = tf.variable(
            tf.randomNormal([h * dk, d]).mul(tf.scalar(Math.sqrt(2 / (h * dk))))
        );
        this.attentionWeights['b_O'] = tf.variable(tf.zeros([d]));

        // 수용체별 쿼리 벡터 (학습 가능)
        this.receptorWeights = [];
        for (let r = 0; r < this.numReceptors; r++) {
            this.receptorWeights.push(
                tf.variable(tf.randomNormal([1, d]).mul(tf.scalar(0.1)))
            );
        }

        console.log(`[CompetitiveBinding] Attention weights initialized (${h} heads, d_k=${dk})`);
    }

    // ==========================================================
    // Scaled Dot-Product Attention
    // ==========================================================
    scaledDotProductAttention(Q, K, V) {
        return tf.tidy(() => {
            const dk = Q.shape[Q.shape.length - 1];
            const scale = tf.scalar(Math.sqrt(dk));

            // QK^T / √d_k
            const scores = tf.matMul(Q, K, false, true).div(scale); // [N, N] or [1, N]

            // Softmax
            const weights = tf.softmax(scores, -1);

            // Weighted sum of V
            const output = tf.matMul(weights, V); // [N, dk] or [1, dk]

            return { output, weights };
        });
    }

    // ==========================================================
    // Multi-Head Attention
    // ==========================================================
    multiHeadAttention(X, queryOverride = null) {
        return tf.tidy(() => {
            const headOutputs = [];
            const allWeights = [];

            for (let i = 0; i < this.numHeads; i++) {
                const W_Q = this.attentionWeights[`W_Q_${i}`];
                const W_K = this.attentionWeights[`W_K_${i}`];
                const W_V = this.attentionWeights[`W_V_${i}`];

                // Q, K, V 프로젝션
                const Q = queryOverride
                    ? tf.matMul(queryOverride, W_Q) // [1, dk]
                    : tf.matMul(X, W_Q); // [N, dk]
                const K = tf.matMul(X, W_K); // [N, dk]
                const V = tf.matMul(X, W_V); // [N, dk]

                const { output, weights } = this.scaledDotProductAttention(Q, K, V);
                headOutputs.push(output);
                allWeights.push(weights);
            }

            // 헤드 Concat + 출력 프로젝션
            const concatenated = tf.concat(headOutputs, -1); // [N, numHeads*dk] or [1, ...]
            const W_O = this.attentionWeights['W_O'];
            const b_O = this.attentionWeights['b_O'];
            const projected = tf.matMul(concatenated, W_O).add(b_O); // [N, d]

            return { output: projected, attentionWeights: allWeights };
        });
    }

    // ==========================================================
    // 분자 임베딩 (입력 변환)
    // ==========================================================
    buildModel() {
        this.initAttentionWeights();

        // 분자 벡터 → 임베딩 공간 프로젝션
        this.projectionModel = tf.sequential();
        this.projectionModel.add(tf.layers.dense({
            inputShape: [29],  // molecular-engine fullMoleculeVector
            units: this.moleculeDim,
            activation: 'relu'
        }));
        this.projectionModel.add(tf.layers.dense({
            units: this.moleculeDim,
            activation: 'tanh'  // [-1, 1] 정규화
        }));

        // 결합 예측 모델 (attention 결과 → 냄새 강도)
        this.bindingModel = tf.sequential();
        this.bindingModel.add(tf.layers.dense({
            inputShape: [this.moleculeDim],
            units: 32,
            activation: 'relu'
        }));
        this.bindingModel.add(tf.layers.dense({
            units: this.numReceptors,
            activation: 'sigmoid'
        }));
        this.bindingModel.compile({
            optimizer: tf.train.adam(0.001),
            loss: 'meanSquaredError'
        });

        console.log('[CompetitiveBinding] Models built');
    }

    async train(moleculeDB, epochs = 30, onProgress = null) {
        if (!this.attentionWeights) this.buildModel();

        const molecules = moleculeDB.getAll ? moleculeDB.getAll() : moleculeDB;
        if (molecules.length === 0) return;

        // 학습 데이터: 개별 분자 → 수용체 활성화 패턴
        const inputs = [];
        const targets = [];

        for (const mol of molecules) {
            const vec = moleculeDB.fullMoleculeVector
                ? moleculeDB.fullMoleculeVector(mol)
                : new Array(29).fill(0);
            inputs.push(vec);

            // 수용체 활성화 할당 (odor_labels와 receptor sensitivity 매칭)
            const activation = this.receptorProfiles.map(r => {
                const labels = mol.odor_labels || [];
                const match = r.sensitivity.filter(s => labels.includes(s)).length;
                return Math.min(1.0, match / r.sensitivity.length + Math.random() * 0.05);
            });
            targets.push(activation);
        }

        const xs = tf.tensor2d(inputs);
        const ys = tf.tensor2d(targets);

        await this.bindingModel.fit(xs, ys, {
            epochs,
            batchSize: 8,
            callbacks: {
                onEpochEnd: (e, logs) => {
                    if (onProgress) onProgress(e + 1, epochs, logs.loss);
                }
            }
        });

        xs.dispose();
        ys.dispose();
        this.trained = true;
        console.log('[CompetitiveBinding] Training complete');
    }

    // ==========================================================
    // 혼합물 시뮬레이션 (진짜 어텐션 기반)
    // ==========================================================
    simulateBinding(molecules, moleculeDB) {
        if (!this.attentionWeights) this.buildModel();

        return tf.tidy(() => {
            const N = molecules.length;
            if (N === 0) return null;

            // 1. 분자 벡터 → 임베딩
            const molVectors = molecules.map(mol =>
                moleculeDB?.fullMoleculeVector
                    ? moleculeDB.fullMoleculeVector(mol)
                    : new Array(29).fill(Math.random() * 0.5)
            );
            const X_raw = tf.tensor2d(molVectors); // [N, 29]
            const X = this.projectionModel.predict(X_raw); // [N, d]

            // 2. Self-Attention: 분자 간 상호작용 학습
            const { output: attended, attentionWeights: selfAttnWeights } = this.multiHeadAttention(X);

            // 3. 수용체별 Cross-Attention: 각 수용체가 분자 혼합물에 주의
            const receptorResults = [];
            for (let r = 0; r < this.numReceptors; r++) {
                const receptorQuery = this.receptorWeights[r]; // [1, d]

                const { output: rOutput, attentionWeights: rWeights } =
                    this.multiHeadAttention(attended, receptorQuery);

                // 수용체 활성화 강도
                const activation = rOutput.abs().mean().dataSync()[0];

                // 각 분자에 대한 어텐션 점수 (첫 번째 헤드)
                const moleculeBindings = rWeights[0].dataSync();

                receptorResults.push({
                    receptorName: this.receptorProfiles[r].name,
                    totalActivation: Math.min(1.0, activation),
                    moleculeBindings: Array.from(moleculeBindings).slice(0, N),
                    dominantMolecule: Array.from(moleculeBindings).slice(0, N)
                        .indexOf(Math.max(...Array.from(moleculeBindings).slice(0, N)))
                });
            }

            // 4. 상호작용 분석 (self-attention 가중치에서)
            const interactions = this._analyzeInteractions(selfAttnWeights, molecules, N);

            // 5. 결과 요약
            const summary = this._generateSummary(receptorResults, interactions, molecules);

            return {
                receptorResults,
                interactions,
                summary,
                numMolecules: N,
                selfAttentionWeights: selfAttnWeights[0].dataSync()
            };
        });
    }

    // 어텐션 가중치 기반 상호작용 분석
    _analyzeInteractions(attentionWeights, molecules, N) {
        const interactions = [];
        const headWeights = attentionWeights[0].dataSync();

        for (let i = 0; i < N; i++) {
            for (let j = i + 1; j < N; j++) {
                // 양방향 어텐션 평균
                const attnIJ = headWeights[i * N + j] || 0;
                const attnJI = headWeights[j * N + i] || 0;
                const mutualAttention = (attnIJ + attnJI) / 2;

                // 상호작용 유형 판단
                let type = 'neutral';
                if (mutualAttention > 0.3) {
                    type = 'synergy'; // 서로 강하게 주의 → 시너지
                } else if (mutualAttention < 0.1 && N > 2) {
                    type = 'masking'; // 매우 낮은 관심 → 마스킹
                }

                interactions.push({
                    moleculeA: molecules[i]?.name || `mol_${i}`,
                    moleculeB: molecules[j]?.name || `mol_${j}`,
                    mutualAttention,
                    type,
                    attnAtoB: attnIJ,
                    attnBtoA: attnJI
                });
            }
        }

        return interactions;
    }

    // 결과 요약 생성
    _generateSummary(receptorResults, interactions, molecules) {
        const summary = [];

        // 가장 활성화된 수용체
        const maxReceptor = receptorResults.reduce((max, r) =>
            r.totalActivation > max.totalActivation ? r : max, receptorResults[0]);
        summary.push(`가장 강한 수용체 반응: ${maxReceptor.receptorName} (${(maxReceptor.totalActivation * 100).toFixed(0)}%)`);

        // 지배적 분자
        if (maxReceptor.dominantMolecule < molecules.length) {
            const dom = molecules[maxReceptor.dominantMolecule];
            summary.push(`지배적 분자: ${dom?.name || 'unknown'}`);
        }

        // 시너지 상호작용
        const synergies = interactions.filter(i => i.type === 'synergy');
        if (synergies.length > 0) {
            const best = synergies[0];
            summary.push(`✨ 시너지: ${best.moleculeA} ↔ ${best.moleculeB} (상호 어텐션 ${(best.mutualAttention * 100).toFixed(0)}%)`);
        }

        // 마스킹 상호작용
        const maskings = interactions.filter(i => i.type === 'masking');
        if (maskings.length > 0) {
            summary.push(`🚫 마스킹 효과: ${maskings.length}건 감지`);
        }

        return summary;
    }

    // ==========================================================
    // 혼합물 냄새 벡터 계산 (진짜 어텐션 가중치 기반)
    // ==========================================================
    computeMixtureOdor(molecules, moleculeDB) {
        if (!this.trained || !this.bindingModel) return null;

        return tf.tidy(() => {
            const N = molecules.length;
            const molVectors = molecules.map(mol =>
                moleculeDB?.fullMoleculeVector
                    ? moleculeDB.fullMoleculeVector(mol)
                    : new Array(29).fill(0)
            );
            const X_raw = tf.tensor2d(molVectors);
            const X = this.projectionModel.predict(X_raw);

            // Self-attention → 가중된 분자 표현
            const { output: attended } = this.multiHeadAttention(X);

            // 평균 풀링 → 혼합물 벡터
            const mixtureVec = attended.mean(0).reshape([1, this.moleculeDim]);

            // 수용체 활성화 예측
            const activation = this.bindingModel.predict(mixtureVec);
            return Array.from(activation.dataSync());
        });
    }
}

export default CompetitiveBinding;

File: server/models/competitive_binding.py
# competitive_binding.py — Pillar 2: nn.MultiheadAttention (PyTorch CUDA)
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np


class MoleculeProjector(nn.Module):
    """분자 벡터 → 임베딩 공간"""
    def __init__(self, input_dim=29, embed_dim=32):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(input_dim, 64),
            nn.ReLU(),
            nn.Linear(64, embed_dim),
            nn.Tanh()
        )
    def forward(self, x):
        return self.net(x)


class CompetitiveBinding:
    """nn.MultiheadAttention 기반 수용체 결합 시뮬레이션"""

    RECEPTOR_PROFILES = [
        {'name': 'OR1: 플로럴', 'sensitivity': ['floral', 'rose', 'jasmine', 'violet']},
        {'name': 'OR2: 시트러스', 'sensitivity': ['citrus', 'lemon', 'orange', 'bergamot']},
        {'name': 'OR3: 우디', 'sensitivity': ['woody', 'cedar', 'sandalwood', 'vetiver']},
        {'name': 'OR4: 스파이시', 'sensitivity': ['spicy', 'clove', 'cinnamon', 'pepper']},
        {'name': 'OR5: 스위트', 'sensitivity': ['sweet', 'vanilla', 'caramel', 'honey']},
        {'name': 'OR6: 프레시', 'sensitivity': ['fresh', 'clean', 'aquatic', 'ozonic']},
        {'name': 'OR7: 그린', 'sensitivity': ['green', 'leafy', 'herbal', 'grassy']},
        {'name': 'OR8: 머스크', 'sensitivity': ['musk', 'amber', 'warm', 'powdery']},
    ]

    def __init__(self, device='cuda'):
        self.device = torch.device(device if torch.cuda.is_available() else 'cpu')
        self.embed_dim = 32
        self.num_heads = 4
        self.num_receptors = 8

        # PyTorch 네이티브 Multi-Head Attention
        self.self_attention = None
        self.cross_attention = None
        self.projector = None
        self.receptor_queries = None
        self.binding_model = None
        self.trained = False

    def build_model(self):
        self.projector = MoleculeProjector(29, self.embed_dim).to(self.device)

        # ★ nn.MultiheadAttention — PyTorch 네이티브 (cuDNN 가속)
        self.self_attention = nn.MultiheadAttention(
            embed_dim=self.embed_dim,
            num_heads=self.num_heads,
            dropout=0.1,
            batch_first=True
        ).to(self.device)

        self.cross_attention = nn.MultiheadAttention(
            embed_dim=self.embed_dim,
            num_heads=self.num_heads,
            dropout=0.1,
            batch_first=True
        ).to(self.device)

        # 학습 가능한 수용체 쿼리 벡터
        self.receptor_queries = nn.Parameter(
            torch.randn(self.num_receptors, 1, self.embed_dim, device=self.device) * 0.1
        )

        self.binding_model = nn.Sequential(
            nn.Linear(self.embed_dim, 32),
            nn.ReLU(),
            nn.Linear(32, self.num_receptors),
            nn.Sigmoid()
        ).to(self.device)

        print(f"[CompetitiveBinding] nn.MultiheadAttention on {self.device}")

    def train(self, molecular_engine, epochs=30, on_progress=None):
        if self.projector is None:
            self.build_model()

        molecules = molecular_engine.get_all()
        if not molecules:
            return

        inputs, targets = [], []
        for mol in molecules:
            vec = self._mol_to_vec(mol)
            inputs.append(vec)
            activation = [
                min(1.0, len(set(r['sensitivity']) & set(mol.get('odor_labels', []))) / len(r['sensitivity']))
                for r in self.RECEPTOR_PROFILES
            ]
            targets.append(activation)

        X = torch.tensor(np.array(inputs), dtype=torch.float32).to(self.device)
        Y = torch.tensor(np.array(targets), dtype=torch.float32).to(self.device)

        # ★ Fix: include self_attention + cross_attention in optimizer
        params = (list(self.projector.parameters()) +
                  list(self.binding_model.parameters()) +
                  list(self.self_attention.parameters()) +
                  list(self.cross_attention.parameters()) +
                  [self.receptor_queries])
        optimizer = torch.optim.Adam(params, lr=0.001)
        criterion = nn.MSELoss()

        dataset = torch.utils.data.TensorDataset(X, Y)
        loader = torch.utils.data.DataLoader(dataset, batch_size=16, shuffle=True)

        for epoch in range(epochs):
            self.projector.train()
            self.binding_model.train()
            self.self_attention.train()
            epoch_loss = 0.0

            for bx, by in loader:
                optimizer.zero_grad()
                # ★ Fix: per-molecule prediction instead of global mean collapse
                embedded = self.projector(bx)  # [B, 32]
                # Self-attention over batch as sequence: [1, B, 32]
                x_seq = embedded.unsqueeze(0)  # [1, B, 32]
                attended, _ = self.self_attention(x_seq, x_seq, x_seq)
                attended = attended.squeeze(0)  # [B, 32]
                out = self.binding_model(attended)  # [B, 8]
                loss = criterion(out, by)
                loss.backward()
                optimizer.step()
                epoch_loss += loss.item()

            if on_progress:
                on_progress(epoch + 1, epochs, epoch_loss / max(len(loader), 1))

        self.trained = True

    @torch.no_grad()
    def simulate_binding(self, molecules, molecular_engine):
        if self.projector is None:
            self.build_model()

        N = len(molecules)
        if N == 0:
            return None

        # 1. 분자 임베딩
        vecs = [self._mol_to_vec(mol) for mol in molecules]
        X = torch.tensor(np.array(vecs), dtype=torch.float32).to(self.device)
        X_embed = self.projector(X)  # [N, 32]
        X_seq = X_embed.unsqueeze(0)  # [1, N, 32]

        # 2. Self-Attention (분자 간 상호작용)
        self.self_attention.eval()
        attended, self_attn_weights = self.self_attention(X_seq, X_seq, X_seq)
        # self_attn_weights: [1, N, N]

        # 3. 수용체별 Cross-Attention
        self.cross_attention.eval()
        receptor_results = []
        for r in range(self.num_receptors):
            query = self.receptor_queries[r].unsqueeze(0)  # [1, 1, 32]
            _, cross_weights = self.cross_attention(query, attended, attended)
            # cross_weights: [1, 1, N]
            bindings = cross_weights.squeeze().cpu().numpy()
            bindings = bindings[:N] if len(bindings.shape) > 0 else [float(bindings)]

            activation = float(attended.abs().mean())
            receptor_results.append({
                'receptorName': self.RECEPTOR_PROFILES[r]['name'],
                'totalActivation': min(1.0, activation),
                'moleculeBindings': [float(b) for b in bindings],
                'dominantMolecule': int(np.argmax(bindings)) if len(bindings) > 0 else 0
            })

        # 4. 상호작용 분석
        attn = self_attn_weights.squeeze(0).cpu().numpy()  # [N, N]
        interactions = []
        for i in range(N):
            for j in range(i + 1, N):
                mutual = (attn[i][j] + attn[j][i]) / 2
                itype = 'synergy' if mutual > 0.3 else ('masking' if mutual < 0.1 and N > 2 else 'neutral')
                interactions.append({
                    'moleculeA': molecules[i].get('name', f'mol_{i}'),
                    'moleculeB': molecules[j].get('name', f'mol_{j}'),
                    'mutualAttention': float(mutual),
                    'type': itype
                })

        return {
            'receptorResults': receptor_results,
            'interactions': interactions,
            'numMolecules': N
        }

    def _mol_to_vec(self, mol):
        """분자 → 29차원 벡터"""
        labels = mol.get('odor_labels', [])
        all_labels = ['floral', 'citrus', 'woody', 'spicy', 'sweet', 'fresh', 'green',
                      'warm', 'musk', 'fruity', 'rose', 'jasmine', 'cedar', 'vanilla',
                      'amber', 'clean', 'smoky', 'powdery', 'aquatic', 'herbal']
        label_vec = [1.0 if l in labels else 0.0 for l in all_labels]
        props = [
            mol.get('molecular_weight', 200) / 500,
            mol.get('logP', 2) / 10,
            mol.get('vapor_pressure', 1) / 100,
            mol.get('boiling_point', 200) / 500,
            mol.get('hbd', 1) / 5,
            mol.get('hba', 2) / 10,
            mol.get('tpsa', 40) / 200,
            mol.get('rotatable_bonds', 3) / 15,
            mol.get('aromatic_rings', 1) / 5,
        ]
        return label_vec + props
